File: src/surface/hl-gl/tests/fixtures/gl_blend_pipeline.c
/* REAL GL FIXED-FUNCTION BLEND — a real EGL + GLES2 offscreen program that proves the GPU's fixed-function
 * framebuffer blend (glEnable(GL_BLEND) + glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)) genuinely
 * composites, rasterized on lavapipe. This SUPERSEDES the fragment-shader-composite workaround in
 * gl_blend.c: the executor now honors the protocol blend field (commit ce66ccba) and the GL driver already
 * lowers glBlendFunc (service/frame.rs blend_factor_wire), so the composite is done by the blend unit, not
 * by arithmetic in the shader.
 *
 * The scene over a BLACK clear is three vertical strips, all driven by the SAME 50%-alpha GREEN fragment
 * output (0,1,0,0.5) over an opaque RED background — the ONLY difference is the blend enable:
 *   LEFT  third  : just the opaque RED background          -> (255,0,0,255)
 *   MIDDLE third : GREEN a=0.5 drawn with GL_BLEND DISABLED -> OVERWRITES to the raw output (0,255,0,128)
 *   RIGHT third  : GREEN a=0.5 drawn with GL_BLEND ENABLED  -> src-over COMPOSITE over red = (128,128,0,191)
 * The middle vs right strips are the proof: identical geometry + identical fragment output, yet one
 * OVERWRITES and one COMPOSITES — only a live framebuffer-blend unit can produce that difference. src-over:
 *   rgb = fg.rgb*fg.a + dst.rgb*(1-fg.a) = green*0.5 + red*0.5 = (0.5,0.5,0)   -> (128,128,0)
 *   a   = fg.a*fg.a   + dst.a*(1-fg.a)   = 0.5*0.5 + 1.0*0.5   = 0.75          -> 191
 *
 * No hl-specific calls, no vendor headers. Prints "GL_BLEND_PIPELINE_OK" on success. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef int32_t EGLint;
typedef unsigned int EGLBoolean, EGLenum, GLenum, GLbitfield, GLuint;
typedef int GLint, GLsizei;
typedef float GLfloat;
typedef char GLchar;
typedef void *EGLDisplay, *EGLConfig, *EGLContext, *EGLSurface, *EGLNativeDisplayType;

extern EGLDisplay eglGetDisplay(EGLNativeDisplayType);
extern EGLBoolean eglInitialize(EGLDisplay, EGLint *, EGLint *);
extern EGLBoolean eglChooseConfig(EGLDisplay, const EGLint *, EGLConfig *, EGLint, EGLint *);
extern EGLBoolean eglBindAPI(EGLenum);
extern EGLContext eglCreateContext(EGLDisplay, EGLConfig, EGLContext, const EGLint *);
extern EGLSurface eglCreatePbufferSurface(EGLDisplay, EGLConfig, const EGLint *);
extern EGLBoolean eglMakeCurrent(EGLDisplay, EGLSurface, EGLSurface, EGLContext);

extern GLuint glCreateShader(GLenum);
extern void glShaderSource(GLuint, GLsizei, const GLchar *const *, const GLint *);
extern void glCompileShader(GLuint);
extern GLuint glCreateProgram(void);
extern void glAttachShader(GLuint, GLuint);
extern void glLinkProgram(GLuint);
extern void glUseProgram(GLuint);
extern GLint glGetAttribLocation(GLuint, const GLchar *);
extern GLint glGetUniformLocation(GLuint, const GLchar *);
extern void glUniform4f(GLint, GLfloat, GLfloat, GLfloat, GLfloat);
extern void glGenBuffers(GLsizei, GLuint *);
extern void glBindBuffer(GLenum, GLuint);
extern void glBufferData(GLenum, long, const void *, GLenum);
extern void glVertexAttribPointer(GLuint, GLint, GLenum, unsigned char, GLsizei, const void *);
extern void glEnableVertexAttribArray(GLuint);
extern void glEnable(GLenum);
extern void glDisable(GLenum);
extern void glBlendFunc(GLenum, GLenum);
extern void glClearColor(GLfloat, GLfloat, GLfloat, GLfloat);
extern void glClear(GLbitfield);
extern void glViewport(GLint, GLint, GLsizei, GLsizei);
extern void glDrawArrays(GLenum, GLint, GLsizei);
extern void glReadPixels(GLint, GLint, GLsizei, GLsizei, GLenum, GLenum, void *);
extern GLenum glGetError(void);

#define EGL_OPENGL_ES_API 0x30A0
#define GL_VERTEX_SHADER 0x8B31
#define GL_FRAGMENT_SHADER 0x8B30
#define GL_ARRAY_BUFFER 0x8892
#define GL_STATIC_DRAW 0x88E4
#define GL_FLOAT 0x1406
#define GL_TRIANGLES 0x0004
#define GL_COLOR_BUFFER_BIT 0x4000
#define GL_RGBA 0x1908
#define GL_UNSIGNED_BYTE 0x1401
#define GL_BLEND 0x0BE2
#define GL_SRC_ALPHA 0x0302
#define GL_ONE_MINUS_SRC_ALPHA 0x0303
#define GL_NO_ERROR 0

#define W 64
#define H 64

static const char *VS =
    "attribute vec2 aPos;\n"
    "void main() { gl_Position = vec4(aPos, 0.0, 1.0); }\n";
static const char *FS =
    "precision mediump float;\n"
    "uniform vec4 uColor;\n"
    "void main() { gl_FragColor = uColor; }\n";

static int near(unsigned char a, unsigned char b) {
    int d = (int)a - (int)b;
    return (d < 0 ? -d : d) <= 4;
}

static GLuint make_vbo(const float *v, long n) {
    GLuint b;
    glGenBuffers(1, &b);
    glBindBuffer(GL_ARRAY_BUFFER, b);
    glBufferData(GL_ARRAY_BUFFER, n, v, GL_STATIC_DRAW);
    return b;
}

int main(void) {
    setbuf(stdout, NULL);

    EGLDisplay dpy = eglGetDisplay(0);
    if (!eglInitialize(dpy, 0, 0)) { fprintf(stderr, "eglInitialize failed\n"); return 1; }
    eglBindAPI(EGL_OPENGL_ES_API);
    EGLConfig cfg;
    EGLint num = 0;
    eglChooseConfig(dpy, 0, &cfg, 1, &num);
    EGLContext ctx = eglCreateContext(dpy, cfg, 0, 0);
    if (!ctx) { fprintf(stderr, "eglCreateContext failed\n"); return 1; }
    EGLSurface surf = eglCreatePbufferSurface(dpy, cfg, 0);
    if (!surf) { fprintf(stderr, "eglCreatePbufferSurface failed\n"); return 1; }
    if (!eglMakeCurrent(dpy, surf, surf, ctx)) { fprintf(stderr, "eglMakeCurrent failed\n"); return 1; }
    glViewport(0, 0, W, H);

    GLuint vs = glCreateShader(GL_VERTEX_SHADER);
    glShaderSource(vs, 1, &VS, 0);
    glCompileShader(vs);
    GLuint fs = glCreateShader(GL_FRAGMENT_SHADER);
    glShaderSource(fs, 1, &FS, 0);
    glCompileShader(fs);
    GLuint prog = glCreateProgram();
    glAttachShader(prog, vs);
    glAttachShader(prog, fs);
    glLinkProgram(prog);
    glUseProgram(prog);
    GLint aPos = glGetAttribLocation(prog, "aPos");
    if (aPos < 0) aPos = 0;
    GLint locColor = glGetUniformLocation(prog, "uColor");

    /* Fullscreen bg; middle third x in [-1/3, 1/3]; right third x in [1/3, 1]. */
    float full[12]   = { -1,-1,  1,-1,  -1,1,  1,-1,  1,1,  -1,1 };
    float mid[12]    = { -0.333f,-1,  0.333f,-1,  -0.333f,1,  0.333f,-1,  0.333f,1,  -0.333f,1 };
    float right[12]  = {  0.333f,-1,  1,-1,  0.333f,1,  1,-1,  1,1,  0.333f,1 };
    GLuint vboFull = make_vbo(full, sizeof(full));
    GLuint vboMid = make_vbo(mid, sizeof(mid));
    GLuint vboRight = make_vbo(right, sizeof(right));

    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
    glClear(GL_COLOR_BUFFER_BIT);

    /* Draw 1: opaque RED background over the whole frame — blend disabled. */
    glDisable(GL_BLEND);
    glBindBuffer(GL_ARRAY_BUFFER, vboFull);
    glVertexAttribPointer((GLuint)aPos, 2, GL_FLOAT, 0, 8, 0);
    glEnableVertexAttribArray((GLuint)aPos);
    glUniform4f(locColor, 1.0f, 0.0f, 0.0f, 1.0f);
    glDrawArrays(GL_TRIANGLES, 0, 6);

    /* Draw 2: GREEN a=0.5 over the MIDDLE third with blend DISABLED — OVERWRITES to the raw (0,255,0,128). */
    glDisable(GL_BLEND);
    glBindBuffer(GL_ARRAY_BUFFER, vboMid);
    glVertexAttribPointer((GLuint)aPos, 2, GL_FLOAT, 0, 8, 0);
    glEnableVertexAttribArray((GLuint)aPos);
    glUniform4f(locColor, 0.0f, 1.0f, 0.0f, 0.5f);
    glDrawArrays(GL_TRIANGLES, 0, 6);

    /* Draw 3: GREEN a=0.5 over the RIGHT third with blend ENABLED — src-over COMPOSITE over red = (128,128,0,191). */
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    glBindBuffer(GL_ARRAY_BUFFER, vboRight);
    glVertexAttribPointer((GLuint)aPos, 2, GL_FLOAT, 0, 8, 0);
    glEnableVertexAttribArray((GLuint)aPos);
    glUniform4f(locColor, 0.0f, 1.0f, 0.0f, 0.5f);
    glDrawArrays(GL_TRIANGLES, 0, 6);

    unsigned char *px = malloc(W * H * 4);
    memset(px, 0xAB, W * H * 4);
    glReadPixels(0, 0, W, H, GL_RGBA, GL_UNSIGNED_BYTE, px);
    GLenum e = glGetError();
    if (e != GL_NO_ERROR) { printf("GL_BLEND_PIPELINE_READBACK_FAILED err=0x%x\n", e); free(px); return 2; }

#define AT(x, y) (&px[(((y) * W) + (x)) * 4])
    unsigned char *bg = AT(10, 32);   /* left third: opaque RED bg only    -> (255,0,0,255)   */
    unsigned char *ov = AT(32, 32);   /* middle: blend OFF overwrite       -> (0,255,0,128)   */
    unsigned char *cp = AT(54, 32);   /* right: blend ON src-over composite -> (128,128,0,191) */
    printf("GL_BLEND_PIPELINE_BG: %u %u %u %u\n", bg[0], bg[1], bg[2], bg[3]);
    printf("GL_BLEND_PIPELINE_OVERWRITE: %u %u %u %u\n", ov[0], ov[1], ov[2], ov[3]);
    printf("GL_BLEND_PIPELINE_COMPOSITE: %u %u %u %u\n", cp[0], cp[1], cp[2], cp[3]);

    int bg_ok = near(bg[0], 255) && near(bg[1], 0) && near(bg[2], 0) && near(bg[3], 255);
    int ov_ok = near(ov[0], 0) && near(ov[1], 255) && near(ov[2], 0) && near(ov[3], 128);
    int cp_ok = near(cp[0], 128) && near(cp[1], 128) && near(cp[2], 0) && near(cp[3], 191);
    if (bg_ok && ov_ok && cp_ok) {
        printf("GL_BLEND_PIPELINE_OK\n");
        free(px);
        return 0;
    }
    printf("GL_BLEND_PIPELINE_WRONG bg=(%u %u %u %u want 255 0 0 255) overwrite=(%u %u %u %u want 0 255 0 128) "
           "composite=(%u %u %u %u want 128 128 0 191) — fixed-function blend did not composite\n",
           bg[0], bg[1], bg[2], bg[3], ov[0], ov[1], ov[2], ov[3], cp[0], cp[1], cp[2], cp[3]);
    free(px);
    return 3;
}

File: src/apps/husklet/tests/gl_blend_pipeline.rs
//! REAL GL FIXED-FUNCTION BLEND — a real EGL + GLES2 offscreen program that proves the GPU blend unit
//! (glEnable(GL_BLEND) + glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)) genuinely composites on
//! lavapipe. This SUPERSEDES the fragment-shader-composite workaround in gl_blend.rs: the wgpu executor now
//! honors the protocol blend field (commit ce66ccba) and the GL driver already lowers glBlendFunc, so the
//! composite is done by the blend unit — not by arithmetic in the shader.
//!
//! Three vertical strips over a BLACK clear, all driven by the SAME 50%-alpha GREEN output (0,1,0,0.5) over
//! an opaque RED background — the ONLY difference between the middle and right strips is the blend enable:
//!   LEFT   -> opaque RED background only              (255,0,0,255)
//!   MIDDLE -> GREEN a=0.5, blend DISABLED -> OVERWRITE (0,255,0,128)
//!   RIGHT  -> GREEN a=0.5, blend ENABLED  -> COMPOSITE (128,128,0,191)
//! Identical geometry + identical fragment output yet one overwrites and one composites: only a live
//! framebuffer-blend unit produces that. Asserts the exact colors from BOTH the guest glReadPixels and an
//! independent host read, and writes /tmp/hl-demo/gl_blend_pipeline.png.

use std::process::Command;

mod common;
use common::wgpu::WgpuExecutorServer;
use common::{bgra_to_rgba, demo_png_dir, staged_dir, write_png};

const W: u32 = 64;
const H: u32 = 64;

#[test]
fn real_gles2_fixed_function_blend_composites_on_lavapipe() {
    let gl_dir = staged_dir("gl");
    assert!(
        gl_dir.join("libEGL.so.1").exists(),
        "staged libEGL.so.1 missing at {gl_dir:?}"
    );
    assert!(
        gl_dir.join("libGLESv2.so.2").exists(),
        "staged libGLESv2.so.2 missing at {gl_dir:?}"
    );

    let manifest = env!("CARGO_MANIFEST_DIR");
    let out_dir = std::env::temp_dir().join(format!("hl-demo-glblendpipe-{}", std::process::id()));
    std::fs::create_dir_all(&out_dir).unwrap();
    let bin = out_dir.join("gl_blend_pipeline");

    let compile = Command::new("gcc")
        .arg(format!(
            "{manifest}/../../surface/hl-gl/tests/fixtures/gl_blend_pipeline.c"
        ))
        .arg(format!("-L{}", gl_dir.display()))
        .args(["-lEGL", "-lGLESv2"])
        .arg("-o")
        .arg(&bin)
        .output()
        .expect("spawn gcc");
    assert!(
        compile.status.success(),
        "gcc failed:\n{}",
        String::from_utf8_lossy(&compile.stderr)
    );

    let exec = WgpuExecutorServer::start("glblendpipe");
    let adapter = exec.adapter_name();
    assert!(
        adapter.to_lowercase().contains("llvmpipe") || adapter.to_lowercase().contains("lavapipe"),
        "must rasterize on the software Vulkan device, got {adapter:?}"
    );

    let run = Command::new(&bin)
        .env("LD_LIBRARY_PATH", &gl_dir)
        .env("EGL_PLATFORM", "surfaceless")
        .env("HL_GPU_EXEC", exec.sock())
        .env("HL_GL_SURFACE_W", W.to_string())
        .env("HL_GL_SURFACE_H", H.to_string())
        .output()
        .expect("spawn gl_blend_pipeline");
    let stdout = String::from_utf8_lossy(&run.stdout);
    let stderr = String::from_utf8_lossy(&run.stderr);
    eprintln!("--- gl_blend_pipeline stdout ---\n{stdout}\n--- stderr ---\n{stderr}");

    assert!(
        run.status.success(),
        "guest exited non-zero (code {:?})",
        run.status.code()
    );
    assert!(
        stdout.contains("GL_BLEND_PIPELINE_OK"),
        "guest asserted the fixed-function composite over the socket"
    );
    assert!(
        exec.submit_count() > 0,
        "guest submitted batches to the host executor"
    );

    // Host read + PNG. BGRA: bg RED (255,0,0)->(0,0,255); overwrite GREEN (0,255,0)->(0,255,0);
    // composite (128,128,0)->(0,128,128).
    let cap = exec.captured();
    let px = cap
        .rgba8_texture(W, H)
        .expect("no 64x64 target captured off the host executor");
    let near = |a: u8, b: u8| (a as i32 - b as i32).abs() <= 4;
    let at = |x: u32, y: u32| {
        let i = ((y * W + x) * 4) as usize;
        (px[i], px[i + 1], px[i + 2], px[i + 3]) // (B,G,R,A)
    };

    write_png(
        &demo_png_dir().join("gl_blend_pipeline.png"),
        W,
        H,
        &bgra_to_rgba(px),
    );

    let (b, g, r, a) = at(10, 32); // left: opaque RED bg
    assert!(
        near(r, 255) && near(g, 0) && near(b, 0) && near(a, 255),
        "left third must be opaque RED bg (255,0,0,255), got RGBA ({r},{g},{b},{a})"
    );
    let (b, g, r, a) = at(32, 32); // middle: blend OFF -> raw output overwrites
    assert!(
        near(r, 0) && near(g, 255) && near(b, 0) && near(a, 128),
        "middle third (blend DISABLED) must OVERWRITE to the raw output (0,255,0,128), got RGBA ({r},{g},{b},{a})"
    );
    let (b, g, r, a) = at(54, 32); // right: blend ON -> src-over composite
    assert!(
        near(r, 128) && near(g, 128) && near(b, 0) && near(a, 191),
        "right third (blend ENABLED) must src-over COMPOSITE to (128,128,0,191), got RGBA ({r},{g},{b},{a})"
    );

    // The composited olive must genuinely cover the right third — a real per-pixel blend, not a stray texel.
    let composite = px
        .chunks_exact(4)
        .filter(|t| near(t[2], 128) && near(t[1], 128) && near(t[0], 0) && near(t[3], 191))
        .count();
    assert!(
        composite > (W * H / 8) as usize,
        "the composited color must cover the right third, only {composite} matched"
    );

    let _ = std::fs::remove_dir_all(&out_dir);
}
